crates/contract/src/instance.rs

```diff
@@ -54,25 +59,31 @@ impl<P> ContractInstance<P> {
     }
 }
 
-impl<P: Clone> ContractInstance<&P> {
+impl<N, T, P: Clone> ContractInstance<N, T, &P> {
     /// Clones the provider and returns a new contract instance with the cloned provider.
     #[inline]
-    pub fn with_cloned_provider(self) -> ContractInstance<P> {
+    pub fn with_cloned_provider(self) -> ContractInstance<N, T, P> {
         ContractInstance {
             address: self.address,
             provider: self.provider.clone(),
             interface: self.interface,
+            transport: PhantomData,
+            network: PhantomData,
         }
     }
 }
 
-impl<P: TempProvider> ContractInstance<P> {
+impl<N: Network, T: Transport + Clone, P: Provider<N, T>> ContractInstance<N, T, P> {
     /// Returns a transaction builder for the provided function name.
     ///
     /// If there are multiple functions with the same name due to overloading, consider using
     /// the [`ContractInstance::function_from_selector`] method instead, since this will use the
     /// first match.
-    pub fn function(&self, name: &str, args: &[DynSolValue]) -> Result<CallBuilder<&P, Function>> {
+    pub fn function(
+        &self,
+        name: &str,
+        args: &[DynSolValue],
```

Comment:
Need follow-up to make this work ideally without the user having to do `DynSolValue::Address(addr)` etc. for each argument, ideally they can do `contract.function("foo", (a, b, c))` or similar instead of `contract.function("foo", &[DynSolValue::(...), DynSolValue::(...), ...)]`